### PR TITLE
Enable building with gcc-14

### PR DESCRIPTION
**Pull request by am11**

### CODE REVIEW

**Review by jkoritzinsky — APPROVED** [review]

File: src/coreclr/pal/inc/pal.h [hunk 1]
```diff
@@ -3371,8 +3371,9 @@ EXTERN_C PALIMPORT inline RETURN_TYPE PALAPI METHOD_DECL        \
 /* Function multiversioning will never inline a method that is  \
    marked such. However, just to make sure that we don't see    \
    surprises, explicitely mark them as noinline. */             \
-__attribute__((target("lse")))  __attribute__((noinline))       \
-EXTERN_C PALIMPORT inline RETURN_TYPE PALAPI Lse_##METHOD_DECL  \
+EXTERN_C PALIMPORT inline RETURN_TYPE PALAPI                    \
+__attribute__((target("+lse")))  __attribute__((noinline))      \
+PALAPI Lse_##METHOD_DECL                                        \
```
Comment (am11) [inline src/coreclr/pal/inc/pal.h:3376]:
```suggestion
Lse_##METHOD_DECL                                               \
```

**Comment by am11** [comment]
Updated gcc pipeline. gcc14 was picked up in the CI:

```
  -- The C compiler identification is GNU 14.1.0
  -- The CXX compiler identification is GNU 14.1.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/gcc-14 - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/g++-14 - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
```

**Comment by kunalspathak** [comment]
@am11 - did you run into any compilation issues with gcc-14 that does not repro on gcc-12. E.g. https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/706791/logs/284. 

It is not liking the presence of if/else inside `constexpr` method. If I remove the if-else, things work fine. Any thoughts?

```c++
    static constexpr regMaskTP CreateFromRegNum(const regNumber reg, regMaskSmall mask)
    {
#ifdef HAS_MORE_THAN_64_REGISTERS
        if (reg < 64)
        {
        }
        else
        {
        }
        return regMaskTP(mask, RBM_NONE);
#else
        return regMaskTP(mask, RBM_NONE);
#endif
    }
```

**Comment by am11** [comment]
It's pedantically requiring single return for std=c++11 conformance:

```c++
static constexpr regMaskTP CreateFromRegNum(regNumber reg, regMaskSmall mask)
{
#ifdef HAS_MORE_THAN_64_REGISTERS
    return (reg < 64) ? regMaskTP(mask, RBM_NONE) : regMaskTP(RBM_NONE, mask);
#else
    return regMaskTP(mask, RBM_NONE);
#endif
}
```

(with std=c++14, if/else works)

**Comment by kunalspathak** [comment]
> (with std=c++14, if/else works)

hhm, how can i make my code work then in current scenario where we are on gcc-14 and in std=c++11 conformance?

**Comment by am11** [comment]
Have you tried the code provided above?

**Comment by kunalspathak** [comment]
> Have you tried the code provided above?

sorry, earlier I missed that piece and was just staring at c++14 and c++11 conformance. Your suggestion works :)

**Pull request merged**